src/query.rs

```diff
@@ -762,6 +828,195 @@ macro_rules! tuple_impl {
+pub struct PreparedQuery<Q: Query> {
+    memo: (usize, u64),
+    state: Box<[(usize, <Q::Fetch as Fetch<'static>>::State)]>,
+    _query: PhantomData<Q>,
+}
+
+impl<Q: Query> PreparedQuery<Q> {
+    #[cold]
+    fn new(world: &World) -> Self {
+        let memo = world.memo();
+
+        let state = world
+            .archetypes()
+            .enumerate()
+            .filter_map(|(idx, x)| Q::Fetch::prepare(x).map(|state| (idx, state)))
+            .collect();
+
+        Self {
+            memo,
+            state,
+            _query: PhantomData,
+        }
+    }
+
+    /// Borrow the query and the [world][World] as a prerequisite for execution
+    ///
+    /// This will panic if it would violate an existing unique reference
+    /// or construct an invalid unique reference.
+    pub fn borrow<'q>(&'q mut self, world: &'q World) -> PreparedQueryBorrow<'q, Q> {
+        if self.memo != world.memo() {
+            *self = Self::new(world);
+        }
+
+        let meta = world.entities_meta();
+        let archetypes = world.archetypes_inner();
+
+        PreparedQueryBorrow::new(meta, archetypes, &*self.state)
+    }
+
+    /// Execute the query using exclusive access to the [world][World]
+    ///
+    /// This avoids the cost of the dynamic borrow checking.
+    /// that is required by [`borrow`][Self::borrow]
+    pub fn iter_mut<'q>(&'q mut self, world: &'q mut World) -> PreparedQueryIter<'q, Q> {
+        assert_borrow::<Q>();
+
+        if self.memo != world.memo() {
+            *self = Self::new(world);
+        }
+
+        let meta = world.entities_meta();
+        let archetypes = world.archetypes_inner();
+
+        let state: &'q [(usize, <Q::Fetch as Fetch<'q>>::State)] =
+            unsafe { mem::transmute(&*self.state) };
+
+        unsafe { PreparedQueryIter::new(meta, archetypes, state.iter()) }
+    }
+}
+
+/// Represents a combined borrow of both [prepared query][PreparedQuery] and [world][World]
+pub struct PreparedQueryBorrow<'q, Q: Query> {
+    meta: &'q [EntityMeta],
+    archetypes: &'q [Archetype],
+    state: &'q [(usize, <Q::Fetch as Fetch<'static>>::State)],
+    active: bool,
+}
+
+impl<'q, Q: Query> PreparedQueryBorrow<'q, Q> {
+    fn new(
+        meta: &'q [EntityMeta],
+        archetypes: &'q [Archetype],
+        state: &'q [(usize, <Q::Fetch as Fetch<'static>>::State)],
+    ) -> Self {
+        for (idx, state) in state {
+            Q::Fetch::borrow(&archetypes[*idx], *state);
+        }
+
+        Self {
+            meta,
+            archetypes,
+            state,
+            active: false,
+        }
+    }
+
+    /// Execute the prepared query
+    ///
+    /// Must be called only once per prepared query.
+    // The lifetime narrowing here is required for soundness.
+    pub fn iter<'i>(&'i mut self) -> PreparedQueryIter<'i, Q> {
+        if self.active {
+            panic!("borrow already active");
+        }
+        self.active = true;
```

Comment:
Since we're not doing the deferred borrow thing that immediate queries use, could we relax this restriction?

Reply:
Yes, I think you are right. As the lifetime of the returned references is bound to the borrow on `PreparedQueryBorrow` and not on `PreparedQuery` and `World`, I cannot create aliasing mutable references by iterating the same borrow twice.

However, what I do not understand now is why a second call to `QueryBorrow::borrow` is a panic instead of a no-op as `QueryBorrow::iter{,_batched}` have the same lifetime narrowing? (And the transform of an already borrowed query via `with/without` will always release fewer and never more borrows, i.e. it will only leak borrows which I can achieve by forgetting the whole `QueryBorrow` anyway.)

Reply:
I think it would be sound to modify `QueryBorrow` to no-op, yeah. No need to address that in the scope of this PR though.